nydusify: introduce chunkdict generate subcommand

newthifans opened the pull request:
###Notice
Please take a look and give me some code suggestions in third commit, thanks!

### Brief introduction
Add the function of storing multiple nydus images into the database for generating chunkdict
Implement the command “nydusify chunkdict generate --sources”
by invoking the command “nydus-image chunkdict save --bootstrap” .

 ### **Basic usage**
nydusify chunkdict generate --sources localhost:5000/ubuntu:nydus_latest,localhost:5000/ubuntu:nydus_latest_v1

### Details
1. Write command “nydusify chunkdict generate” entries
2. Create a “generater” folder to implement the function
3. Store the image in the registry with slices and Store them in the database in turn
4. Parse nydus image and create a directory to hold corresponding bootstrap
5. Pull nydus image bootstrap and store  corresponding directory
6. Invoke  the command “nydus-image chunkdict save --bootstrap”

This PR is related to #1305 

###Notice
This PR only modifies 6 files which is located at “contrib/nydusify”

cslinwang committed:
nydus-image: Store chunk and blob metadata

Add functionality to store chunk and blob metadata
from nydus source images.
Use the 'nydus-image chunkdict save' command
with the '--bootstrap' option followed by the
path to the nydus bootstrap file (e.g., '~/output/nydus_bootstrap')

Signed-off-by: Lin Wang <l.wang@mail.dlut.edu.cn>

Desiki-high commented:
/retest

imeoer commented:
@newthifans Please rename the PR title to make it meaningful, and fix the code conflicts, thanks!

newthifans commented:
> @newthifans Please rename the PR title to make it meaningful, and fix the code conflicts, thanks!

Thanks，I have renamed the PR title. About code conflicts, this PR's first two commits came from PR #1305. The conflicts and bugs of the first two commits will be fixed in #1305. The third commit is my code. Please take a look and give me some code suggestions, thanks!

imeoer commented on Cargo.toml:
```diff
@@ -57,6 +57,7 @@ openssl = { version = "0.10.48", features = ["vendored"] }
 # pin openssl-src to bring in fix for https://rustsec.org/advisories/RUSTSEC-2022-0032
 #openssl-src = { version = "111.22" }
 
+
```
Why is there a such commit?

imeoer commented on contrib/nydusify/cmd/nydusify.go:
```diff
@@ -629,6 +630,75 @@ func main() {
+						&cli.StringFlag{
+							Name:    "work-dir",
+							Value:   "./output",
+							Usage:   "Working directory for image verification",
```
Why `image verification`?

imeoer commented on contrib/nydusify/cmd/nydusify.go:
```diff
@@ -629,6 +630,75 @@ func main() {
+							EnvVars: []string{"NYDUS_IMAGE"},
+						},
+						&cli.StringFlag{
+							Name:    "nydusd",
```
Do we depend on this component?

imeoer commented on contrib/nydusify/go.mod:
```diff
@@ -29,6 +28,8 @@ require (
 	lukechampine.com/blake3 v1.1.5
 )
 
+require github.com/goharbor/acceleration-service v0.2.4
```
Can we move it to the above?

imeoer commented on contrib/nydusify/pkg/generater/bootstrap.go:
```diff
@@ -0,0 +1,35 @@
+// Copyright 2020 Ant Group. All rights reserved.
```
Copyright 2023 Nydus Developers. All rights reserved.

imeoer commented on contrib/nydusify/pkg/generater/generater.go:
```diff
@@ -0,0 +1,115 @@
+package generater
```
Can we rename the package to `chunkdict`?

imeoer commented on contrib/nydusify/pkg/generater/generater.go:
```diff
@@ -0,0 +1,115 @@
+// by Nydus-image to save image chunk and blob into database.
+type Generater struct {
+	Opt
+	sourceSlice  []string
```
`sources`

imeoer commented on contrib/nydusify/pkg/generater/generater.go:
```diff
@@ -0,0 +1,115 @@
+				generater.sourceParser[index].Remote.MaybeWithHTTP(err)
+			}
+		}
+		generater.generate(ctx, index)
```
The error should be handled.

imeoer commented on contrib/nydusify/pkg/generater/generater.go:
```diff
@@ -0,0 +1,115 @@
+	if err != nil {
+		return errors.Wrap(err, "parse Nydus image")
+	}
+	//create a directory to store the image bootstrap
```
Add a space after `//`.

imeoer commented on contrib/nydusify/pkg/generater/generater.go:
```diff
@@ -0,0 +1,115 @@
+	//create a directory to store the image bootstrap
+	nydusImageName := generater.sourceSlice[index][strings.LastIndex(generater.sourceSlice[index], "/")+1:]
+	folderPath := filepath.Join(generater.WorkDir, nydusImageName)
+	_, err = os.ReadDir(folderPath)
```
Should use `os.Stat`?

imeoer commented on contrib/nydusify/pkg/generater/generater.go:
```diff
@@ -0,0 +1,115 @@
+	_, err = os.ReadDir(folderPath)
+	if err == nil {
+		if err := os.RemoveAll(folderPath); err != nil {
+			return errors.Wrap(err, "error delete directory")
```
return errors.Wrap(err, "remove work directory")

imeoer commented on contrib/nydusify/pkg/generater/generater.go:
```diff
@@ -0,0 +1,115 @@
+			return errors.Wrap(err, "error delete directory")
+		}
+		if err := os.Mkdir(folderPath, fs.ModePerm); err != nil {
+			return errors.Wrap(err, "error creat directory")
```
return errors.Wrap(err, "create work directory")

imeoer commented on contrib/nydusify/pkg/generater/tool/builder.go:
```diff
@@ -0,0 +1,50 @@
+	}
+}
+
+// Save calls `nydus-image save` to parse Nydus bootstrap
```
`nydus-image chunkdict generate`?

newthifans replied:
`nydus-image chunkdict save`  achieve function "store nydus image into database", "nydus-image chunkdict generate" achieve algorithm afterwards.

imeoer commented on contrib/nydusify/pkg/generater/output.go:
```diff
@@ -0,0 +1,62 @@
+			return errors.Wrap(err, "output nydus index file")
+		}
+	}
+	if sourceParsed.NydusImage != nil {
```
Should give a error massage if the source is not a nydus image.

imeoer commented:
The PR title should be `nydusify: introduce chunkdict generate subcommand`.

hangvane commented on contrib/nydusify/cmd/nydusify.go:
```diff
@@ -629,6 +630,75 @@ func main() {
+			Subcommands: []*cli.Command{
+				{
+					Name:  "generate",
+					Usage: "Multi-image save chunk info to a database",
```
"Save chunk info of multiple images to a database" ?

hangvane commented on contrib/nydusify/pkg/generater/generater.go:
```diff
@@ -0,0 +1,115 @@
+)
+
+// Opt defines Chunkdict generate options.
+// Note: sources is one and more Nydus image references.
```
one or more?

hangvane commented on contrib/nydusify/pkg/generater/generater.go:
```diff
@@ -0,0 +1,115 @@
+	sourceParser []*parser.Parser
+}
+
+// New creates Generater instance, sourceSlice is multiple Nydus images reference list.
```
Please consider moving the comment for the variable `sourceSlice` to the line above it.

hangvane commented on contrib/nydusify/pkg/generater/generater.go:
```diff
@@ -0,0 +1,115 @@
+	return generater, nil
+}
+
+// "Chunkdict Generate" deposit multiple Nydus images to the database in turn
```
Could you please explain `in turn`?

hangvane commented on src/bin/nydus-image/deduplicate.rs:
```diff
@@ -0,0 +1,239 @@
+use nydus_storage::meta::format_blob_features;
+use rusqlite::{params, Connection};
+pub struct Deduplicate {
+    sb: RafsSuper,
```
Consider using a clearer variable name.

jiangliu commented on Cargo.toml:
```diff
@@ -58,13 +59,17 @@ openssl = { version = "0.10.48", features = ["vendored"] }
 nydus-api = { version = "0.2.2", path = "api", features = ["handler"] }
 nydus-rafs = { version = "0.2.2", path = "rafs", features = ["builder"] }
-nydus-service = { version = "0.2.0", path = "service", features = ["block-device"] }
+nydus-service = { version = "0.2.0", path = "service", features = [
```
Unrelated change？

newthifans replied:
This is the change made by the first two commits from **#1305**. I don't understand this part, I work on the third commit,  just calling the subcommands `nydus-image` chunkdict save ” implemented by the first two commits.

imeoer commented on contrib/nydusify/pkg/chunkdict/generater/bootstrap.go:
```diff
@@ -0,0 +1,36 @@
+}
+
+func (bootstrap *Bootstrap) Save() error {
+	logrus.Infof("Save chunk and blob table into database")
```
Next, we need to call `nydus-image chunkdict generate` to generate the chunk dict image directly, so let's remove the info message.

imeoer commented on contrib/nydusify/pkg/chunkdict/generater/generater.go:
```diff
@@ -0,0 +1,117 @@
+// Note: sources is one or more Nydus image references.
+type Opt struct {
+	WorkDir        string
+	Sources        string
```
Why not `Sources []string`.

imeoer commented on contrib/nydusify/pkg/chunkdict/generater/generater.go:
```diff
@@ -0,0 +1,117 @@
+// "Chunkdict generate" command store multiple Nydus images into database
+// by invoking "nydus-image chunkdict save" command
+// to save chunk and blob information image into database.
+type Generater struct {
```
Generator.

imeoer commented on contrib/nydusify/pkg/chunkdict/generater/generater.go:
```diff
@@ -0,0 +1,117 @@
+	ExpectedArch   string
+}
+
+// "Chunkdict generate" command store multiple Nydus images into database
```
`chunkdict generate` should be used for generating the chunk dict image directly.

imeoer commented on contrib/nydusify/pkg/chunkdict/generater/generater.go:
```diff
@@ -0,0 +1,117 @@
+type Generater struct {
+	Opt
+	// Nydus image references slice.
+	sourcesSlice  []string
```
It's better not to attach a type to a variable name.

imeoer commented on contrib/nydusify/pkg/chunkdict/generater/generater.go:
```diff
@@ -0,0 +1,117 @@
+		sourceParser, err := parser.New(sourcesRemote, opt.ExpectedArch)
+		sourcesParser = append(sourcesParser, sourceParser)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create parser")
```
Please use consistent capitalization in the error message.

imeoer commented on contrib/nydusify/pkg/chunkdict/generater/generater.go:
```diff
@@ -0,0 +1,117 @@
+				generater.sourcesParser[index].Remote.MaybeWithHTTP(err)
+			}
+		}
+		generater.generate(ctx, index)
```
We should not miss any error handling.

imeoer commented on contrib/nydusify/pkg/chunkdict/generater/generater.go:
```diff
@@ -0,0 +1,117 @@
+	return nil
+}
+
+// A Nydus image is stored in the database
```
Can we explain the generate method clearly?

imeoer commented on contrib/nydusify/pkg/chunkdict/generater/generater.go:
```diff
@@ -0,0 +1,117 @@
+	}
+
+	// Create a directory to store the image bootstrap
+	nydusImageName := generater.sourcesSlice[index][strings.LastIndex(generater.sourcesSlice[index], "/")+1:]
```
How can we ensure no have conflict name with a different registry host?

imeoer commented on contrib/nydusify/pkg/chunkdict/generater/generater.go:
```diff
@@ -0,0 +1,117 @@
+	nydusImageName := generater.sourcesSlice[index][strings.LastIndex(generater.sourcesSlice[index], "/")+1:]
+	folderPath := filepath.Join(generater.WorkDir, nydusImageName)
+	_, err = os.Stat(folderPath)
+	if err == nil {
```
Can we use os.MkdirAll here?
`os.MkdirAll`:  `If path is already a directory, MkdirAll does nothing and returns nil.`

imeoer replied:
And we need to clean up the work directory after the generation finish.

Desiki-high commented on contrib/nydusify/pkg/build/builder.go:
```diff
@@ -143,3 +147,24 @@ func (builder *Builder) Run(option BuilderOption) error {
+	cmd.Stdout = builder.stdout
+	cmd.Stderr = builder.stderr
+
+	if err := cmd.Run(); err != nil {
+		return err
+	}
+	return nil
```
Maybe ```return cmd.Run()```

Desiki-high commented on contrib/nydusify/pkg/chunkdict/generator/generator.go:
```diff
@@ -0,0 +1,107 @@
+
+// "Chunkdict Generate" deposit multiple Nydus images into the database one by one.
+func (generator *Generator) Generate(ctx context.Context) error {
+	for index := 0; index < len(generator.sourcesParser); index++ {
+		if err := generator.save(ctx, index); err != nil {
+			if utils.RetryWithHTTP(err) {
+				generator.sourcesParser[index].Remote.MaybeWithHTTP(err)
+			}
+			if err := generator.save(ctx, index); err != nil {
+				return err
+			}
+		}
+	}
```
Maybe use ```for index, sourceParser := range generator.sourcesParser``` is better than ```len(generator.sourcesParser)``` ?

Desiki-high commented on contrib/nydusify/pkg/chunkdict/generator/generator.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+// "Save" function stores information of chunk and blob of a Nydus Image in the database
+func (generator *Generator) save(ctx context.Context, index int) error {
```
Does ```index int``` better than ```sourceParser *parser.Parser```? why we must need the index?

newthifans replied:
`Index int` indicates two  the ordinal numbers of the two arrays `Source` and  `sourceParser`. Using `index` is easier.

newthifans committed:
nydusify: Store chunk and blob metadata of multi-image into datbase

nydusify chunkdict generate --sources
Add the 'nydus-image chunkdict save' command
with the "--sources" followed by the nydus image of registry
(e.g.,'registry.com/busybox:nydus-v1,registry.com/busybox:nydus-v2')

Signed-off-by: Zhao Yuan <1627990440@qq.com>
